File: app/src/main/java/hcm/ditagis/com/cholon/qlsc/async/LoadingDataFeatureAsync.java
package hcm.ditagis.com.cholon.qlsc.async;

import android.app.Activity;
import android.content.Context;
import android.os.AsyncTask;
import android.support.v7.widget.LinearLayoutManager;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.esri.arcgisruntime.data.ArcGISFeature;
import com.esri.arcgisruntime.data.CodedValue;
import com.esri.arcgisruntime.data.CodedValueDomain;
import com.esri.arcgisruntime.data.Field;

import java.util.ArrayList;
import java.util.List;

import hcm.ditagis.com.cholon.qlsc.R;
import hcm.ditagis.com.cholon.qlsc.entities.DApplication;
import hcm.ditagis.com.cholon.qlsc.utities.Constant;

public class LoadingDataFeatureAsync extends AsyncTask<Void, Void, Void> {
    private AsyncResponse mDelegate;
    private Activity mActivity;
    private Context mContext;
    private List<Field> mFields;
    private ArcGISFeature mArcGISFeature;
    private DApplication mApplication;

    public interface AsyncResponse {
        void processFinish(List<View> views);
    }

    public LoadingDataFeatureAsync(Activity activity, Context context, List<Field> fields, AsyncResponse delegate, ArcGISFeature... arcGISFeatures) {
        mActivity = activity;
        mApplication = (DApplication) activity.getApplication();
        mContext = context;
        mFields = fields;
        mDelegate = delegate;
        if (arcGISFeatures != null && arcGISFeatures.length > 0)
            mArcGISFeature = arcGISFeatures[0];
    }

    @Override
    protected Void doInBackground(Void... voids) {
        publishProgress();
        return null;
    }

    @Override
    protected void onProgressUpdate(Void... values) {
        mDelegate.processFinish(loadDataToAdd());
    }

    private List<View> loadDataToAdd() {
        List<View> views = new ArrayList<>();
        LinearLayoutManager layoutManager = new LinearLayoutManager(mContext);
        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        for (Field field : mFields) {
            String name = field.getName();
            if (name.equals(Constant.FieldSuCo.DIA_CHI)
                    || name.equals(Constant.FieldSuCo.NGUOI_PHAN_ANH)
                    || name.equals(Constant.FieldSuCo.SDT_PHAN_ANH)
                    || name.equals(Constant.FieldSuCo.THONG_TIN_PHAN_ANH)
                    || name.equals(Constant.FieldSuCo.GHI_CHU)
                    )
                views.add(getView(field));

        }
        return views;
    }

    private View getView(Field field) {
        LinearLayout layoutView = (LinearLayout) mActivity.getLayoutInflater().inflate(R.layout.item_add_feature, null);
        LinearLayout layoutEditNumber = layoutView.findViewById(R.id.llayout_add_feature_number);
        LinearLayout layoutEditNumberDecimal = layoutView.findViewById(R.id.llayout_add_feature_number_decimal);
        LinearLayout layoutEditSpinner = layoutView.findViewById(R.id.llayout_add_feature_spinner);
        LinearLayout layoutEditText = layoutView.findViewById(R.id.llayout_add_feature_edittext);


        Spinner spin = layoutEditSpinner.findViewById(R.id.spinner_add_spinner_value);
        ArrayAdapter<String> adapter = new ArrayAdapter<>(mContext, android.R.layout.simple_list_item_1, new ArrayList<>());
        spin.setAdapter(adapter);
        Object value = null;
        if (mArcGISFeature != null) {
            value = mArcGISFeature.getAttributes().get(field.getName());
        }
        if (field.getDomain() != null) {
            CodedValueDomain codedValueDomain = (CodedValueDomain) field.getDomain();
            List<String> values = new ArrayList<>();
            values.add(Constant.NULL);
            String selectedValue = null;
            for (CodedValue codedValue : codedValueDomain.getCodedValues()) {
                values.add(codedValue.getName());
                if (value != null && codedValue.getCode().equals(value))
                    selectedValue = codedValue.getName();
            }

            layoutEditNumberDecimal.setVisibility(View.GONE);
            layoutEditSpinner.setVisibility(View.VISIBLE);
            layoutEditText.setVisibility(View.GONE);
            layoutEditNumber.setVisibility(View.GONE);

            TextView textViewSpin = layoutEditSpinner.findViewById(R.id.txt_add_spiner_title);
            textViewSpin.setText(field.getAlias());
            adapter.addAll(values);
            adapter.notifyDataSetChanged();

            for (int i = 0; i < values.size(); i++) {
                if (selectedValue != null && values.get(i).equals(selectedValue)) {
                    spin.setSelection(i);
                    break;
                }
            }
        } else switch (field.getFieldType()) {
            case INTEGER:
            case SHORT:
                layoutEditNumberDecimal.setVisibility(View.GONE);
                layoutEditSpinner.setVisibility(View.GONE);
                layoutEditText.setVisibility(View.GONE);
                layoutEditNumber.setVisibility(View.VISIBLE);

                TextView textViewNumber = layoutEditNumber.findViewById(R.id.txt_add_edit_number_title);
                textViewNumber.setText(field.getAlias());

                if (value != null) {
                    EditText editTextNumber = layoutView.findViewById(R.id.etxt_add_edit_number_value);
                    try {
                        switch (field.getFieldType()) {
                            case INTEGER:
                                editTextNumber.setText(Integer.parseInt(value.toString()));
                                break;
                            case SHORT:
                                editTextNumber.setText(Short.parseShort(value.toString()));
                                break;
                        }
                    } catch (Exception e) {
                        Toast.makeText(mContext, "Có lỗi khi load dữ liệu", Toast.LENGTH_SHORT).show();
                    }
                }
                break;
            case DOUBLE:
            case FLOAT:
                layoutEditNumberDecimal.setVisibility(View.VISIBLE);
                layoutEditSpinner.setVisibility(View.GONE);
                layoutEditText.setVisibility(View.GONE);
                layoutEditNumber.setVisibility(View.GONE);

                TextView textViewNumberDecimal = layoutEditNumberDecimal.findViewById(R.id.txt_add_edit_number_decimal_title);
                textViewNumberDecimal.setText(field.getAlias());

                if (value != null) {
                    EditText editTextNumberDecimal = layoutView.findViewById(R.id.etxt_add_edit_number_decimal_value);
                    try {
                        switch (field.getFieldType()) {
                            case DOUBLE:
                                editTextNumberDecimal.setText(Double.parseDouble(value.toString()) + "");
                                break;
                            case FLOAT:
                                editTextNumberDecimal.setText(Float.parseFloat(value.toString()) + "");
                                break;
                        }
                    } catch (Exception e) {
                        Toast.makeText(mContext, "Có lỗi khi load dữ liệu", Toast.LENGTH_SHORT).show();
                    }
                }
                break;
            case TEXT:
                layoutEditNumberDecimal.setVisibility(View.GONE);
                layoutEditSpinner.setVisibility(View.GONE);
                layoutEditNumber.setVisibility(View.GONE);
                layoutEditText.setVisibility(View.VISIBLE);
                TextView textViewEditText = layoutEditText.findViewById(R.id.txt_add_edit_text_title);
                textViewEditText.setText(field.getAlias());
                if (field.getName().equals(Constant.FieldSuCo.DIA_CHI))
                    value = mApplication.getDiemSuCo().getVitri();
                if (value != null) {
                    EditText editText = layoutView.findViewById(R.id.edit_add_edittext_value);
                    try {
                        editText.setText(value.toString());
                    } catch (Exception e) {
                        Toast.makeText(mContext, "Có lỗi khi load dữ liệu", Toast.LENGTH_SHORT).show();
                    }
                }
                break;
            default:
                layoutEditNumberDecimal.setVisibility(View.GONE);
                layoutEditSpinner.setVisibility(View.GONE);
                layoutEditNumber.setVisibility(View.GONE);
                layoutEditText.setVisibility(View.GONE);
        }
        return layoutView;
    }


}

File: app/src/main/java/hcm/ditagis/com/cholon/qlsc/async/LoadingDataFeatureAsync.kt
package hcm.ditagis.com.cholon.qlsc.async

import android.app.Activity
import android.content.Context
import android.os.AsyncTask
import androidx.recyclerview.widget.LinearLayoutManager
import android.view.View
import android.widget.*
import com.esri.arcgisruntime.data.ArcGISFeature
import com.esri.arcgisruntime.data.CodedValueDomain
import com.esri.arcgisruntime.data.Field
import hcm.ditagis.com.cholon.qlsc.R
import hcm.ditagis.com.cholon.qlsc.entities.DApplication
import hcm.ditagis.com.cholon.qlsc.utities.Constant
import java.util.*

open class LoadingDataFeatureAsync(private val mActivity: Activity, private val mContext: Context, private val mFields: List<Field>, private val mDelegate: AsyncResponse, vararg arcGISFeatures: ArcGISFeature)
    : AsyncTask<Boolean, Boolean, Void>() {
    private var mArcGISFeature: ArcGISFeature? = null
    private val mApplication: DApplication = mActivity.application as DApplication

    interface AsyncResponse {
        fun processFinish(views: List<View>)
    }

    init {
        if (arcGISFeatures != null && arcGISFeatures.size > 0)
            mArcGISFeature = arcGISFeatures[0]
    }

 override fun doInBackground(vararg params: Boolean?): Void? {
        if (params != null && params.size > 0)
            publishProgress(params[0])
        return null
    }

    override fun onProgressUpdate(vararg values: Boolean?) {
        mDelegate.processFinish(loadDataToAdd(values[0]))
    }

    private fun loadDataToAdd(isAdd: Boolean?): List<View> {
        val views = ArrayList<View>()
        val layoutManager = LinearLayoutManager(mContext)
        layoutManager.orientation = LinearLayoutManager.VERTICAL
        val outFields = mApplication.dFeatureLayer!!.getdLayerInfo().outFieldsArr
        var isFoundField = false
        for (field in mFields) {
            val name = field.name
            if (isAdd!!) {
                if (name == Constant.FieldSuCo.DIA_CHI
                        || name == Constant.FieldSuCo.NGUOI_PHAN_ANH
                        || name == Constant.FieldSuCo.SDT_PHAN_ANH
                        || name == Constant.FieldSuCo.THONG_TIN_PHAN_ANH
                        || name == Constant.FieldSuCo.GHI_CHU)
                    views.add(getView(field))

            } else {
                if (outFields.size > 0 && outFields[0] != "*") {
                    for (s in outFields)
                        if (s == field.name) {
                            isFoundField = true
                            break
                        }
                    if (isFoundField) {
                        isFoundField = false
                    } else
                        continue
                }
                if (name == Constant.FieldSuCo.ID_SUCO
                        || name == Constant.Field.OBJECTID
                        || name == Constant.FieldSuCo.MA_DMA
                        || name == Constant.FieldSuCo.MA_DUONG
                        || name == Constant.FieldSuCo.MA_PHUONG
                        || name == Constant.FieldSuCo.MA_QUAN
                        || name == Constant.FieldSuCo.SDT_PHAN_ANH
                        || name == Constant.FieldSuCo.NGUOI_PHAN_ANH)
                    continue
                views.add(getView(field))
            }
        }
        return views
    }

    private fun getView(field: Field): View {
        val layoutView = mActivity.layoutInflater.inflate(R.layout.item_add_feature, null) as LinearLayout
        val layoutEditNumber = layoutView.findViewById<LinearLayout>(R.id.llayout_add_feature_number)
        val layoutEditNumberDecimal = layoutView.findViewById<LinearLayout>(R.id.llayout_add_feature_number_decimal)
        val layoutEditSpinner = layoutView.findViewById<LinearLayout>(R.id.llayout_add_feature_spinner)
        val layoutEditText = layoutView.findViewById<LinearLayout>(R.id.llayout_add_feature_edittext)


        val spin = layoutEditSpinner.findViewById<Spinner>(R.id.spinner_add_spinner_value)
        val adapter = ArrayAdapter(mContext, android.R.layout.simple_list_item_1, ArrayList<String>())
        spin.adapter = adapter
        var value: Any? = null
        if (mArcGISFeature != null) {
            value = mArcGISFeature!!.attributes[field.name]
        }
        if (field.domain != null) {
            val codedValueDomain = field.domain as CodedValueDomain
            val values = ArrayList<String>()
            values.add(Constant.NULL)
            var selectedValue: String? = null
            for (codedValue in codedValueDomain.codedValues) {
                values.add(codedValue.name)
                if (value != null && codedValue.code == value)
                    selectedValue = codedValue.name
            }

            layoutEditNumberDecimal.visibility = View.GONE
            layoutEditSpinner.visibility = View.VISIBLE
            layoutEditText.visibility = View.GONE
            layoutEditNumber.visibility = View.GONE

            val textViewSpin = layoutEditSpinner.findViewById<TextView>(R.id.txt_add_spiner_title)
            textViewSpin.text = field.alias
            adapter.addAll(values)
            adapter.notifyDataSetChanged()

            for (i in values.indices) {
                if (selectedValue != null && values[i] == selectedValue) {
                    spin.setSelection(i)
                    break
                }
            }
        } else
            when (field.fieldType) {
                Field.Type.INTEGER, Field.Type.SHORT -> {
                    layoutEditNumberDecimal.visibility = View.GONE
                    layoutEditSpinner.visibility = View.GONE
                    layoutEditText.visibility = View.GONE
                    layoutEditNumber.visibility = View.VISIBLE

                    val textViewNumber = layoutEditNumber.findViewById<TextView>(R.id.txt_add_edit_number_title)
                    textViewNumber.text = field.alias

                    if (value != null) {
                        val editTextNumber = layoutView.findViewById<EditText>(R.id.etxt_add_edit_number_value)
                        try {
                            when (field.fieldType) {
                                Field.Type.INTEGER -> editTextNumber.setText(Integer.parseInt(value.toString()))
                                Field.Type.SHORT -> editTextNumber.setText(java.lang.Short.parseShort(value.toString()).toInt())
                            }
                        } catch (e: Exception) {
                            Toast.makeText(mContext, "Có lỗi khi load dữ liệu", Toast.LENGTH_SHORT).show()
                        }

                    }
                }
                Field.Type.DOUBLE, Field.Type.FLOAT -> {
                    layoutEditNumberDecimal.visibility = View.VISIBLE
                    layoutEditSpinner.visibility = View.GONE
                    layoutEditText.visibility = View.GONE
                    layoutEditNumber.visibility = View.GONE

                    val textViewNumberDecimal = layoutEditNumberDecimal.findViewById<TextView>(R.id.txt_add_edit_number_decimal_title)
                    textViewNumberDecimal.text = field.alias

                    if (value != null) {
                        val editTextNumberDecimal = layoutView.findViewById<EditText>(R.id.etxt_add_edit_number_decimal_value)
                        try {
                            when (field.fieldType) {
                                Field.Type.DOUBLE -> editTextNumberDecimal.setText(java.lang.Double.parseDouble(value.toString()).toString() + "")
                                Field.Type.FLOAT -> editTextNumberDecimal.setText(java.lang.Float.parseFloat(value.toString()).toString() + "")
                            }
                        } catch (e: Exception) {
                            Toast.makeText(mContext, "Có lỗi khi load dữ liệu", Toast.LENGTH_SHORT).show()
                        }

                    }
                }
                Field.Type.TEXT -> {
                    layoutEditNumberDecimal.visibility = View.GONE
                    layoutEditSpinner.visibility = View.GONE
                    layoutEditNumber.visibility = View.GONE
                    layoutEditText.visibility = View.VISIBLE
                    val textViewEditText = layoutEditText.findViewById<TextView>(R.id.txt_add_edit_text_title)
                    textViewEditText.text = field.alias
                    if (field.name == Constant.FieldSuCo.DIA_CHI)
                        value = mApplication.diemSuCo.vitri
                    if (value != null) {
                        val editText = layoutView.findViewById<EditText>(R.id.edit_add_edittext_value)
                        try {
                            editText.setText(value.toString())
                        } catch (e: Exception) {
                            Toast.makeText(mContext, "Có lỗi khi load dữ liệu", Toast.LENGTH_SHORT).show()
                        }

                    }
                }
                else -> {
                    layoutEditNumberDecimal.visibility = View.GONE
                    layoutEditSpinner.visibility = View.GONE
                    layoutEditNumber.visibility = View.GONE
                    layoutEditText.visibility = View.GONE
                }
            }
        return layoutView
    }


}